testing/trino-tests/src/test/java/io/trino/execution/TestQueryTracker.java

```diff
@@ -0,0 +1,98 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.Session;
+import io.trino.connector.MockConnectorFactory;
+import io.trino.spi.Plugin;
+import io.trino.spi.connector.ColumnMetadata;
+import io.trino.spi.connector.ConnectorFactory;
+import io.trino.testing.DistributedQueryRunner;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import static io.trino.SystemSessionProperties.QUERY_MAX_PLANNING_TIME;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static java.lang.Integer.MAX_VALUE;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+// Tests need to finish before strict timeouts. Any background work
+// may make them flaky
+@Test(singleThreaded = true)
+public class TestQueryTracker
+{
+    private DistributedQueryRunner queryRunner;
+
+    private volatile boolean interrupted;
+
+    @BeforeMethod
+    public void setup()
+            throws Exception
+    {
+        Session defaultSession = testSessionBuilder()
+                .setCatalog("mock")
+                .setSchema("default")
+                .setSystemProperty(QUERY_MAX_PLANNING_TIME, "5s")
+                .build();
+
+        queryRunner = DistributedQueryRunner
+                .builder(defaultSession)
+                .build();
+        queryRunner.installPlugin(new Plugin()
+        {
+            @Override
+            public Iterable<ConnectorFactory> getConnectorFactories()
+            {
+                return ImmutableList.of(MockConnectorFactory.builder()
+                        .withGetColumns(ignored -> ImmutableList.of(new ColumnMetadata("col", VARCHAR)))
+                        // Apply filter happens inside optimizer so this should model most blocking tasks in planning phase
+                        .withApplyFilter((ignored1, ignored2, ignored3) -> freeze())
+                        .build());
+            }
+        });
+        queryRunner.createCatalog("mock", "mock");
+    }
+
+    @AfterMethod(alwaysRun = true)
+    public void tearDown()
+    {
+        queryRunner.close();
+        queryRunner = null;
+    }
+
+    @Test(timeOut = 10_000)
+    public void testInterruptApplyFilter()
+    {
+        assertThatThrownBy(() -> queryRunner.execute("SELECT * FROM t1 WHERE col = 'abc'"))
+                .hasMessageContaining("Query exceeded the maximum planning time limit of 5.00s");
+        assertThat(interrupted).isTrue();
```

Comment:
This is racy, planning thread might have hiccup and set `interrupted = true` later than `queryRunner.execute` returns.

Use https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CountDownLatch.html instead.